lib/pos_epoch_complete_hook.go

```diff
@@ -0,0 +1,373 @@
+}
+
+// simulatePrevEpochEntry simulates the block range for the previous epoch given the current epoch's initial
+// block height and epoch number. The view and timestamp for the simulated epoch are left empty since they can't
+// be easily simulated, so DO NOT USE CreatedAtBlockTimestampNanoSecs or InitialView from the returned EpochEntry.
+func (bav *UtxoView) simulatePrevEpochEntry(currentEpochNumber uint64, currentEpochInitialBlockHeight uint64) (*EpochEntry, error) {
+	if currentEpochNumber == 0 {
+		return nil, errors.New("simulatePrevEpochEntry: currentEpochNumber is 0, this should never happen")
+	}
+	snapshotAtEpochNumber, err := bav.ComputeSnapshotEpochNumberForEpoch(currentEpochNumber - 1)
+	if err != nil {
+		return nil, errors.Wrapf(err, "simulatePrevEpochEntry: problem computing snapshot epoch number: ")
+	}
+	snapshotGlobalParamsEntry, err := bav.GetSnapshotGlobalParamsEntryByEpochNumber(snapshotAtEpochNumber)
+	if err != nil {
+		return nil, errors.Wrapf(err, "simulatePrevEpochEntry: problem retrieving snapshot global params entry: ")
+	}
+
+	// Calculate the PrevEpoch's InitialBlockHeight.
+	prevEpochInitialBlockHeight, err := SafeUint64().Sub(currentEpochInitialBlockHeight, snapshotGlobalParamsEntry.EpochDurationNumBlocks)
+	if err != nil {
+		return nil, errors.Wrapf(err, "simulatePrevEpochEntry: problem calculating PrevEpochInitialBlockHeight: ")
+	}
+	prevEpochEntry := &EpochEntry{
+		EpochNumber:                     currentEpochNumber - 1,
+		InitialBlockHeight:              prevEpochInitialBlockHeight,
+		InitialView:                     0,
+		FinalBlockHeight:                currentEpochInitialBlockHeight - 1,
+		InitialLeaderIndexOffset:        0,
+		CreatedAtBlockTimestampNanoSecs: 0,
+	}
+	return prevEpochEntry, nil
+}
+
+// Updates the currentEpochEntry to the next epoch's.
+func (bav *UtxoView) runEpochCompleteEpochRollover(
+	epochNumber uint64,
+	blockHeight uint64,
+	view uint64,
+	previousView uint64,
+	blockTimestampNanoSecs int64,
+) error {
+	nextEpochEntry, err := bav.computeNextEpochEntry(epochNumber, blockHeight, view, previousView, blockTimestampNanoSecs)
+	if err != nil {
+		return errors.Wrap(err, "runEpochCompleteEpochRollover: ")
+	}
+	bav._setCurrentEpochEntry(nextEpochEntry)
+
+	return nil
+}
+
+func (bav *UtxoView) generateAndSnapshotValidatorSet(epochNumber uint64, validatorSet []*ValidatorEntry,
+) {
+	// Snapshot the current top n active validators as the validator set.
+	for _, validatorEntry := range validatorSet {
+		bav._setSnapshotValidatorSetEntry(validatorEntry, epochNumber)
+	}
+
+	// Snapshot the current validator set's total stake. Note, the validator set is already filtered to the top n
+	// active validators for the epoch. The total stake is the sum of all of the active validators' stakes.
+	validatorSetTotalStakeAmountNanos := SumValidatorEntriesTotalStakeAmountNanos(validatorSet)
+	bav._setSnapshotValidatorSetTotalStakeAmountNanos(validatorSetTotalStakeAmountNanos, epochNumber)
```

Comment:
Wow, surprising that this remained unused. The data is correctly populated in the UtxoView and in the DB. Looks like everywhere we check if a block has supermajority vote, we manually sum up the validators' stakes. This is all done in the `consensus` package so it makes sense why this value is never read from there.

I don't feel strongly about removing this since it requires a DB change. WDYT?

Reply:
resolved in #1318